File: gradle/packages/cli-function/src/main/kotlin/com/prog/cli/function/Number.kt
package com.prog.cli.function

import com.prog.numbers.utils.operations.FactorialUtils.factorial
import com.prog.numbers.utils.operations.FibonacciUtils.fibonacci

object Number {
    /**
     * # `main`
     *
     * This function is the entry point of the application.
     *
     * __Arguments:__
     *
     * * `args: Array<String>`
     */
    @JvmStatic
    fun main(args: Array<String>) {
        println("===================================================")
        printNumberCalculations()
        println("===================================================")
    }

    /**
     * # `getUserNumber`
     *
     * This function prompts the user to enter a number and returns the input.
     *
     * __Return:__
     *
     * * `Int` - The user's number
     *
     * __Example:__
     *
     * ```kotlin
     * import com.prog.cli.function.Number.getUserNumber
     * val number = getUserNumber()
     * println("Number : $number")
     * ```
     *
     * __Output:__
     *
     * ```kotlin
     * "Enter a number : 5"
     * "Number : 5"
     * ```
     */
    private fun getUserNumber(): Int {
        print("Enter a number : ")
        return readLine()?.trim()?.toIntOrNull() ?: 0
    }

    /**
     * # `printNumberCalculations`
     *
     * This function prints the Fibonacci and Factorial of the user's number by
     * using the `fibonacci` and `factorial` functions from the `FibonacciUtils`
     */
    fun printNumberCalculations() {
        val number = getUserNumber()
        println("Fib[$number] = ${fibonacci(number)}")
        println("Fact[$number] = ${factorial(number)}")
    }
}

File: haskell/src/CLI/Function/Number.hs
module CLI.Function.Number (printNumberCalculations) where

import System.IO (hFlush, stdout)
import Text.Read (readMaybe)

import Numbers.Utils.Operations.FactorialUtils (factorial)
import Numbers.Utils.Operations.FibonacciUtils (fibonacci)

-- |
-- = @getUserNumber@
--
-- The function @getUserNumber@ is used to get the user number.
--
-- __Return:__
--
-- * @IO Integer@ - The user number.
--
-- __Example:__
--
-- @
-- import CLI.Function.Number (getUserNumber)
-- number <- getUserNumber
-- putStrLn $ "Number : " ++ show number
-- @
--
-- __Output:__
--
-- @
-- "Enter a number : 10"
-- "Number : 10"
-- @
getUserNumber :: IO Integer
getUserNumber = do
    putStr "Enter a number : "
    hFlush stdout
    maybeNum <- fmap readMaybe getLine
    case maybeNum of
        Just num -> return num
        Nothing -> return 0

-- |
-- = @printNumberCalculations@
--
-- The function @printNumberCalculations@ is used to print the fibonacci and
-- factorial of the user number by calling the @fibonacci@ and @factorial@.
printNumberCalculations :: IO ()
printNumberCalculations = do
    number <- getUserNumber
    putStrLn $ "Fib[" ++ show number ++ "] = " ++ show (fibonacci number)
    putStrLn $ "Fact[" ++ show number ++ "] = " ++ show (factorial number)
